File: integrations/openclaw/__tests__/test_datasetName.ts
/**
 * Regression tests for sanitizeDatasetName() and its integration into
 * datasetNameForScope() — issue #3549.
 *
 * Place this file at:
 *   integrations/openclaw/__tests__/test_datasetName.ts
 *
 * Run with the existing test runner (e.g. npx jest or npm test).
 */

import { sanitizeDatasetName, datasetNameForScope } from "../src/scope.js";
import type { CogneePluginConfig, MemoryScope } from "../src/types.js";

// ---------------------------------------------------------------------------
// Shared fixture — minimal Required<CogneePluginConfig>
// ---------------------------------------------------------------------------
function makeCfg(overrides: Partial<CogneePluginConfig> = {}): Required<CogneePluginConfig> {
  return {
    baseUrl: "http://localhost:8000",
    datasetName: "openclaw",
    searchType: "GRAPH_COMPLETION",
    minScore: 0.3,
    topK: 5,
    requestTimeout: 60_000,
    ingestionTimeout: 300_000,
    deleteMode: "soft",
    maxTokens: 512,
    autoRecall: true,
    autoRecallScope: ["agent", "user", "company"],
    scopeRouting: [],
    defaultScope: "agent" as MemoryScope,
    companyDataset: "",
    userDatasetPrefix: "",
    agentDatasetPrefix: "",
    agentDatasetTemplate: "",
    userId: "user1",
    agentId: "agent1",
    perAgentMemory: false,
    recallInjectionPosition: "system",
    persistSession: false,
    ...overrides,
  } as unknown as Required<CogneePluginConfig>;
}

// ---------------------------------------------------------------------------
// Issue #3549 canonical test cases — identical across all integrations
// ---------------------------------------------------------------------------
const CASES: [string, string, string][] = [
  // [input, fallback, expected]
  ["mydataset",    "fb", "mydataset"],
  ["project_01",   "fb", "project_01"],
  ["company-prod", "fb", "company-prod"],
  ["test.v1",      "fb", "test.v1"],
  ["My Dataset",   "fb", "My_Dataset"],
  ["My@Data",      "fb", "My_Data"],
  ["!!hello!!",    "fb", "hello"],
  ["###",          "fb", "fb"],
  ["........",     "fb", "fb"],
  ["___",          "fb", "fb"],
  ["a".repeat(200), "fb", "a".repeat(120)],
  ["你好 Dataset",  "fb", "Dataset"],
  ["",             "fb", "fb"],
  ["-leading",     "fb", "-leading"],
  [".leading",     "fb", "leading"],
  ["_leading",     "fb", "leading"],
  ["trailing.",    "fb", "trailing"],
];

describe("sanitizeDatasetName", () => {
  test("issue #3549 canonical cases", () => {
    for (const [input, fallback, expected] of CASES) {
      expect(sanitizeDatasetName(input, fallback)).toBe(expected);
    }
  });

  test("max length is 120", () => {
    expect(sanitizeDatasetName("x".repeat(200), "fb")).toHaveLength(120);
  });

  test("fallback fires only when result is empty after sanitization", () => {
    expect(sanitizeDatasetName("ok_name", "fb")).toBe("ok_name");
    expect(sanitizeDatasetName("###",     "fb")).toBe("fb");
  });

  test("null/undefined treated as empty string", () => {
    expect(sanitizeDatasetName(undefined as unknown as string, "fb")).toBe("fb");
    expect(sanitizeDatasetName(null as unknown as string, "fb")).toBe("fb");
  });
});

// ---------------------------------------------------------------------------
// datasetNameForScope — verifies sanitization is applied in every branch
// ---------------------------------------------------------------------------
describe("datasetNameForScope sanitizes composite names", () => {
  test("company scope: bad companyDataset is sanitised", () => {
    const cfg = makeCfg({ companyDataset: "My Company!" });
    expect(datasetNameForScope("company", cfg)).toBe("My_Company");
  });

  test("company scope: bad datasetName propagates and is sanitised", () => {
    const cfg = makeCfg({ datasetName: "my@claw", companyDataset: "" });
    // composite = "my@claw-company" → "my_claw-company"
    expect(datasetNameForScope("company", cfg)).toBe("my_claw-company");
  });

  test("user scope: userId with special chars is sanitised", () => {
    const cfg = makeCfg({ userId: "user@example.com", userDatasetPrefix: "" });
    // "openclaw-user-user@example.com" → "openclaw-user-user_example.com"
    expect(datasetNameForScope("user", cfg)).toBe("openclaw-user-user_example.com");
  });

  test("agent scope: agentDatasetTemplate with special chars is sanitised", () => {
    const cfg = makeCfg({ agentDatasetTemplate: "!!{agentId}!!", agentId: "bot" });
    // "!!bot!!" → "__bot__" → strip → "bot"
    expect(datasetNameForScope("agent", cfg)).toBe("bot");
  });

  test("agent scope: fully-invalid template falls back", () => {
    const cfg = makeCfg({ agentDatasetTemplate: "###", agentId: "bot" });
    expect(datasetNameForScope("agent", cfg)).toBe("openclaw-agent-default");
  });

  test("valid config produces unchanged dataset names", () => {
    const cfg = makeCfg({ datasetName: "openclaw", agentId: "bot" });
    expect(datasetNameForScope("company", cfg)).toBe("openclaw-company");
    expect(datasetNameForScope("user",    cfg)).toBe("openclaw-user-user1");
    expect(datasetNameForScope("agent",   cfg)).toBe("openclaw-agent-bot");
  });
});

File: integrations/hermes-agent/tests/test_dataset_name.py
"""Regression tests for _sanitize_dataset_name in the hermes-agent integration.

Covers every valid/invalid input from issue #3549 and verifies that the
hermes-agent sanitizer produces exactly the expected output.

Run:
    python integrations/hermes-agent/tests/test_dataset_name.py
"""

import pathlib
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parents[1]))

from cognee_integration_hermes.provider import _sanitize_dataset_name  # noqa: E402

DEFAULT = "hermes"

# (input_name, expected_output)
# When expected_output is DEFAULT the sanitizer must fall back to the default.
CASES = [
    # ── valid names — must pass through unchanged ──
    ("mydataset",    "mydataset"),
    ("project_01",   "project_01"),
    ("company-prod", "company-prod"),
    ("test.v1",      "test.v1"),
    # ── invalid names — must be normalised ──
    ("My Dataset",   "My_Dataset"),     # space → _
    ("My@Data",      "My_Data"),        # @ → _
    ("!!hello!!",    "hello"),          # leading/trailing _ stripped
    ("###",          DEFAULT),          # all _ after replace → stripped → empty → fallback
    ("........",     DEFAULT),          # all . stripped → empty → fallback
    ("___",          DEFAULT),          # all _ stripped → empty → fallback
    ("a" * 200,      "a" * 120),        # truncated to 120
    ("你好 Dataset",  "Dataset"),        # non-ASCII → _, then stripped
    # ── edge cases ──
    ("",             DEFAULT),          # empty string → fallback
    ("  ",           DEFAULT),          # spaces only → __ → stripped → fallback
    ("-leading",     "-leading"),       # leading - is NOT stripped (only . and _ are)
    (".leading",     "leading"),        # leading . stripped
    ("_leading",     "leading"),        # leading _ stripped
    ("trailing.",    "trailing"),       # trailing . stripped
    ("a" * 120,      "a" * 120),        # exactly at the limit — unchanged
    ("a" * 121,      "a" * 120),        # one over the limit — truncated
]


def test_sanitize_dataset_name_valid_names() -> None:
    for name, expected in CASES[:4]:
        result = _sanitize_dataset_name(name, DEFAULT)
        assert result == expected, (
            f"[valid] _sanitize_dataset_name({name!r}) = {result!r}, want {expected!r}"
        )


def test_sanitize_dataset_name_invalid_names() -> None:
    for name, expected in CASES[4:]:
        result = _sanitize_dataset_name(name, DEFAULT)
        assert result == expected, (
            f"[invalid] _sanitize_dataset_name({name!r}) = {result!r}, want {expected!r}"
        )


def test_fallback_used_only_when_empty_after_sanitization() -> None:
    """Fallback must NOT fire for names that are non-empty after sanitization."""
    assert _sanitize_dataset_name("ok_name", "fallback") == "ok_name"
    assert _sanitize_dataset_name("###", "fallback") == "fallback"


def test_max_length_is_120() -> None:
    long = "x" * 200
    result = _sanitize_dataset_name(long, DEFAULT)
    assert len(result) == 120, f"expected length 120, got {len(result)}"


if __name__ == "__main__":
    failures = 0
    for _name, _fn in sorted(globals().items()):
        if _name.startswith("test_") and callable(_fn):
            try:
                _fn()
                print("PASS", _name)
            except AssertionError as exc:
                failures += 1
                print("FAIL", _name, "→", exc)
    raise SystemExit(1 if failures else 0)
